src/lib/utils/debounce.js

```diff
@@ -0,0 +1,26 @@
+/**
+ * A simple debounce util for ensuring a function only gets called once during
```

Comment:
Yes but, your function is actually async. What if Algolia takes >=200ms to run?

This should actually look something like...

```js
export const debounce = (func, wait) => {
  let p = Promise.resolve();
  let pendingPromise;
  let pendingArgs;
  return function(...args) {
    if (!pendingPromise) {
      pendingPromise = new Promise((r) => r(func(...pendingArgs)));  // promisify 'func' and ensure latest args are passed
      const safe = pendingPromise.catch(null).then(() => {
        pendingPromise = undefined;
        pendingArgs = undefined;
      });
      const delay = new Promise((r) => setTimeout(r, wait));
      p = p.then(Promise.all([delay, safe]));
    }
    pendingArgs = args;
    return pendingPromise;
  };
}
```

This will allow the function to run immediately (possibly desirable), then wait for it to finish (ignoring any errors) **and** the delay to finish. Alternative behavior may be to always ensure ~200ms between calls (in which case the promises could run serially).

It can possibly be simpler but this is the quickest I could code up. Hopefully you see what I mean.

Reply:
> Yes but, your function is actually async. What if Algolia takes >=200ms to run?

In that scenario I was thinking [the locally cached query would act as a guard](https://github.com/GoogleChrome/web.dev/pull/1633/files#diff-3549ac7324c1b1a52b2f376e7bceaee1R243-R244). So if I type "a" and it takes 200ms to run, but in the meantime I type "bc" and it returns in 100ms, then when "a" returns from algolia, it looks at the current query "abc" which doesn't match "a" so it ignores the response from "a".

Totally possible that I'm missing something though, so happy to chat about this when you're online to make sure I'm understanding things correctly.

Reply:
I might be reading your function incorrectly, but does it prevent querying if a request is currently pending?

For example, if I type "a", it will set `pendingPromise` to wait until the query for "a" has returned. If I type other keys in the meantime, will it continue handing back the promise for "a"?

I'm also a little unclear on what this line is doing:
```js
p = p.then(Promise.all([delay, safe]));
```

Reply:
Ah, yeah, I made a mistake when I wrote that function _late_ last night. It's storing the `pendingArgs` but also sending off the query immediately. Oops.

Maybe:

```js
export const debounce = (func, wait) => {
  let p = Promise.resolve();
  let pending;

  return function(...args) {
    const wasPending = (pending !== undefined);
    pending = args;
    if (wasPending) { return; }

    // Protect from failures in any previous calls, delay for |wait|, and call func with the latest pending args.
    p = p.catch(null).then(async () => {
      await new Promise((r) => setTimeout(r, wait));
      try {
        return await func(...pending);
      } finally {
        pending = undefined;  // future calls will delay for |wait| again
      }
    });
  };
}
```

As I said in the top-level comment, it's also fine to skip this and leave a TODO for now.